net/api/test_server.cc.h

```diff
@@ -9,17 +9,216 @@
+};
+
+struct Routes {
+  struct Helper {
+    struct Impl {
+      explicit Impl(Routes* routes) : routes_(routes) {}
+      ~Impl() {
+        assert(method_ != Method::UNINITIALIZED);
+        assert(!path_.empty());
+        assert(handler_);
+        routes_->AddRoute(method_, path_, handler_);
+      }
+      void SetHandler(std::function<void(Request&&)>&& f) {
+        assert(!handler_);
+        handler_ = f;
+      }
+      void AddConstraint(Method method) {
+        assert(method != Method::UNINITIALIZED);
+        assert(method_ == Method::UNINITIALIZED);
+        method_ = method;
+      }
+      void AddConstraint(const std::string& path) {
+        assert(!path.empty());
+        assert(path_.empty());
+        path_ = path;
+      }
+
+      Routes* routes_;
+      std::function<void(Request&&)> handler_;
+
+      Method method_ = Method::UNINITIALIZED;
+      std::string path_ = "";
+
+      Impl() = delete;
+      Impl(const Impl&) = delete;
+      Impl(Impl&&) = delete;
+      void operator=(const Impl&) = delete;
+      void operator=(Impl&&) = delete;
+    };
+
+    Helper(Routes* routes) : impl_(new Impl(routes)) {}
+    Helper(Helper&& rhs) { impl_.swap(rhs.impl_); }
+    void operator=(std::function<void(Request&&)>&& f) {
+      assert(impl_);
+      impl_->SetHandler(std::forward<std::function<void(Request && )>>(f));
+    }
+    template <typename T>
+    Helper& operator[](T&& constraint) {
+      assert(impl_);
+      impl_->AddConstraint(std::forward<T>(constraint));
+      return *this;
+    }
+
+    std::unique_ptr<Impl> impl_;
+  };
+
+  template <typename T>
+  Helper operator[](T&& constraint) {
+    Helper helper(this);
+    helper[constraint];
+    return helper;
+  }
+
+  void AddRoute(Method method, const std::string& path, std::function<void(Request&&)> handler) {
+    routes_[std::make_pair(method, path)] = handler;
+  }
+
+  void HandleRequest(HTTPServerConnection& connection) const {
+    const HTTPReceivedMessage& message = connection.Message();
+    const Method method = message.Method() == "GET" ? Method::GET : Method::POST;  // TODO(dkorolev).
+    const std::string& path = message.Path();  // TODO(dkorolev): Rename message.Path(), it's incorrect.
+    const url::URL url = url::URL(path);
+    const auto cit = routes_.find(std::make_pair(method, url.path));
+    if (cit != routes_.end()) {
+      Request request{url, connection, message};
+      cit->second(std::forward<Request>(request));
+    } else {
+      ASSERT_TRUE(false) << "Method not implemented.";
+    }
+  }
+
+  std::map<std::pair<Method, std::string>, std::function<void(Request&&)>> routes_;
+};
+
+class Server {
+ public:
+  Server(int port, bool& time_to_terminate, Routes routes)
+      : port_(port),
+        time_to_terminate_(time_to_terminate),
+        routes_(routes),
+        thread_(&Server::Code, this, Socket(port_)) {}
+
+  ~Server() {
+    HTTP(GET(strings::Printf("localhost:%d/killtestserver", port_)));
```

Comment:
True that, although keep in mind that ```::accept()``` is a blocking call in POSIX, and making another server connection is a much easier way to signal the server that is has to be terminated.

Even if it would be an HTTP route at the end, the user will not see it.